MetaMask/metamask-extension
# Add TypeScript to the linting process

This commit allows developers to write TypeScript files and lint them
(either via a language server in their editor of choice or through the
`yarn lint` command).

The new TypeScript configuration as well as the updated ESLint
configuration not only includes support for parsing TypeScript files,
but also provides some compatibility between JavaScript and TypeScript.
That is, it makes it possible for a TypeScript file that imports a
JavaScript file or a JavaScript file that imports a TypeScript file to
be linted.

Note that this commit does not integrate TypeScript into the build
system yet, so we cannot start converting files to TypeScript and
pushing them to the repo until that final step is complete.

---

Manual testing steps:

- Check out this branch locally.
- Run `git cherry-pick 4bde1eb1d0bf0e84d76f83bb0bb065cfc1f45e58`.
  - This commit (which is disconnected from this branch) converts two files to TypeScript, `ui/helpers/utils/tx-helper.ts` and `ui/helpers/utils/storage-helpers.ts`. Both of these files have dependents which are JavaScript files, and `ui/helpers/utils/tx-helper.ts` has dependencies which are JavaScript files.
- Run `yarn setup`.
- Open `ui/helpers/utils/storage-helpers.ts` in your editor. Remove the semicolon from the first line; your editor should tell you (via ESLint) that this is invalid.
- Now open `ui/index.js`. This file imports `tx-helper.ts`. Mouse over `txHelper` in the `import txHelper` line; you should see proper types for this function which come from the TypeScript file.
- Run `yarn lint`. All files should pass.
